mknod /dev/ptmx instead of bind mounting it

jordemort opened the pull request:
This compensates for a problem in versions of systemd older than 219, which causes units with "PrivateDevices=yes" to fail to start with a "226/NAMESPACE" error.

Fixes https://github.com/lxc/lxc/issues/1623

It's my understanding that the original issue was left unfixed because systemd < 219 is relatively ancient, in terms of systemd versions, and that it had been fixed upstream already (https://github.com/systemd/systemd/pull/3254) - however, we have existing container setups based on Debian Jessie, which ships systemd 215, and this is preventing us from migrating to a newer version of LXC.

It looked like the original approach with a symlink was unacceptable because it confused apparmor, wheras the bind mount approach breaks older versions of systemd that didn't handle private /dev correctly. I've replaced the bind mount with a call to `mknod`; in a nutshell, this patch calls `stat` on `/dev/pts/ptmx` and then creates a separate-but-identical device node at '/dev/ptmx`. This works for us; I'm hoping you'll find this somewhat compromised approach acceptable in the name of keeping Jessie-based containers working.

lxc-jenkins commented:
This pull request didn't trigger Jenkins as its author isn't in the whitelist.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

jordemort commented:
Note that Jessie also ships with units that contain `PrivateDevices=yes` (`systemd-hostnamed` is what ended up biting us here, specifically) so it's not simply a matter of "just don't do that there" - this problem breaks the default install of the distribution.

stgraber commented:
Hmm, won't this break nesting and fully unprivileged containers where mknod isn't allowed?

jordemort commented:
@stgraber I'm not sure about nesting but yeah it would definitely break if mknod is not allowed. The symlink fallback code is still there; I could maybe put the bind mount stuff back as an additional fallback to the mknod?

jordemort commented:
I've added the bind mount back as a fallback. So now it:
  1. Tries mknod
  2. Tries bind mount
  3. Tries symlink

I also restructured things a little bit in the mknod stuff to make things more consistent with what was already there, and updated a couple comments.

brauner commented:
You should use am_root in the handler and if you are create a second console device otherwise execute the bind code and then the symlink code. Sorry, on my phone so not really detailed. :)

brauner commented on src/lxc/conf.c:
```diff
@@ -1433,6 +1434,40 @@ static int lxc_setup_devpts(int num_pts)
 
+	/* Read attributes of /dev/pts/ptmx */
+	ret = stat("/dev/pts/ptmx", &ptmx);
+	if (ret) {
```
Please make that `(if ret < 0)`

brauner commented on src/lxc/conf.c:
```diff
@@ -1433,6 +1434,40 @@ static int lxc_setup_devpts(int num_pts)
+	if (ret) {
+		SYSERROR("failed to stat \"/dev/pts/ptmx\"");
+		return -1;
+	} else if (!S_ISCHR(ptmx.st_mode)) {
```
Since you return in the  `if` branch above the `else if` is not really needed. Avoid the indentation and just perform the check unconditionally. Meaning:

```
if (/* char dev check fails */)
       return error;
```

brauner commented on src/lxc/conf.c:
```diff
@@ -1433,6 +1434,40 @@ static int lxc_setup_devpts(int num_pts)
+
+	/* Create doppelganger of /dev/pts/ptmx at /dev/ptmx */
+	ret = mknod("/dev/ptmx", ptmx.st_mode, ptmx.st_rdev);
+	if (!ret) {
```
The handler records - on startup - whether you are root or not. So you should either give this fun an additional arg to inform it whether it is root or not or pass down the handler itself and. Then you can check here and only perform the `mknod()` syscall when you are root. This saves a context switch.

brauner commented on src/lxc/conf.c:
```diff
@@ -1433,6 +1434,40 @@ static int lxc_setup_devpts(int num_pts)
+		/* Our work is done here */
+		DEBUG("made copy of \"/dev/pts/ptmx\" at \"/dev/ptmx\"");
+		return 0;
+	} else {
```
Again, the else is not needed since since you are returning anyway.

jordemort commented:
Thanks for the feedback, @brauner! I'm not sure if I'll have time to get back to it today but I'll be working on this this week.

brauner commented:
@jordemort, np. Thanks for the patch! :)

brauner commented:
Haven't heard back in  while. Please re-open a new PR if still something you need.
